File: deepgoal/core/container.py
from __future__ import annotations

import asyncio
from collections.abc import Callable

from .executor import Executor
from .node import Node, PipelineInput, PipelineOutput, PipelineResult
from .types import EngineOptions

PipelineItem = Node | Executor


def output_to_input(previous: PipelineInput, output: PipelineOutput) -> PipelineInput:
    return PipelineInput(
        primary_path=output.primary_path,
        artifacts=[*previous.artifacts, *output.artifacts],
        metadata={**previous.metadata, **output.metadata},
    )


async def run_pipeline_item(item: PipelineItem, input: PipelineInput, options: EngineOptions) -> PipelineResult:
    if isinstance(item, Node):
        return await item.run(input, options)
    if isinstance(item, Executor):
        return await item.execute(input, options)
    raise TypeError(f"Unsupported pipeline item: {type(item).__name__}")


class SerialNode(Node):
    """Run child items sequentially, passing each output into the next item."""

    def __init__(self, items: list[PipelineItem]) -> None:
        self._items = items

    async def run(self, input: PipelineInput, options: EngineOptions) -> PipelineResult:
        current = input
        result = PipelineResult.ok(output=input.to_output())
        for item in self._items:
            result = await run_pipeline_item(item, current, options)
            if not result.success:
                return result
            current = output_to_input(current, result.output)
        return result


class ParallelNode(Node):
    """Run child items concurrently with the same input and merge their results."""

    def __init__(self, items: list[PipelineItem], merge: Callable[[list[PipelineResult]], PipelineResult]) -> None:
        self._items = items
        self._merge = merge

    async def run(self, input: PipelineInput, options: EngineOptions) -> PipelineResult:
        results = await asyncio.gather(*[run_pipeline_item(item, input, options) for item in self._items])
        return self._merge(list(results))


class LoopNode(Node):
    """Repeat an item until the predicate is satisfied or the item fails."""

    def __init__(self, inner: PipelineItem, until: Callable[[PipelineResult], bool]) -> None:
        self._inner = inner
        self._until = until

    async def run(self, input: PipelineInput, options: EngineOptions) -> PipelineResult:
        current = input
        while True:
            result = await run_pipeline_item(self._inner, current, options)
            if not result.success or self._until(result):
                return result
            current = output_to_input(current, result.output)


class BranchNode(Node):
    """Select a child item at runtime based on the current input."""

    def __init__(self, selector: Callable[[PipelineInput], PipelineItem]) -> None:
        self._selector = selector

    async def run(self, input: PipelineInput, options: EngineOptions) -> PipelineResult:
        item = self._selector(input)
        return await run_pipeline_item(item, input, options)

File: deepgoal-ts/src/core/container.ts
import { Executor } from './executor.js';
import { Node, createPipelineInput, inputToOutput, okResult, type PipelineInput, type PipelineOutput, type PipelineResult } from './node.js';
import type { EngineOptions } from './types/index.js';

export type PipelineItem = Node | Executor;
export type ParallelMerge = (results: PipelineResult[]) => PipelineResult;
export type LoopPredicate = (result: PipelineResult) => boolean;
export type BranchSelector = (input: PipelineInput) => PipelineItem;

export function outputToInput(previous: PipelineInput, output: PipelineOutput): PipelineInput {
  return createPipelineInput({
    primaryPath: output.primaryPath,
    artifacts: [...previous.artifacts, ...output.artifacts],
    metadata: { ...previous.metadata, ...output.metadata },
  });
}

export async function runPipelineItem(item: PipelineItem, input: PipelineInput, options: EngineOptions): Promise<PipelineResult> {
  if (item instanceof Node) {
    return item.run(input, options);
  }
  if (item instanceof Executor) {
    return item.execute(input, options);
  }
  throw new TypeError(`Unsupported pipeline item: ${typeof item}`);
}

export class SerialNode extends Node {
  constructor(private readonly items: PipelineItem[]) {
    super();
  }

  override async run(input: PipelineInput, options: EngineOptions): Promise<PipelineResult> {
    let current = input;
    let result = okResult(inputToOutput(input));

    for (const item of this.items) {
      result = await runPipelineItem(item, current, options);
      if (!result.success) {
        return result;
      }
      current = outputToInput(current, result.output);
    }

    return result;
  }
}

export class ParallelNode extends Node {
  constructor(
    private readonly items: PipelineItem[],
    private readonly merge: ParallelMerge
  ) {
    super();
  }

  override async run(input: PipelineInput, options: EngineOptions): Promise<PipelineResult> {
    const results = await Promise.all(this.items.map(item => runPipelineItem(item, input, options)));
    return this.merge(results);
  }
}

export class LoopNode extends Node {
  constructor(
    private readonly inner: PipelineItem,
    private readonly until: LoopPredicate
  ) {
    super();
  }

  override async run(input: PipelineInput, options: EngineOptions): Promise<PipelineResult> {
    let current = input;

    while (true) {
      const result = await runPipelineItem(this.inner, current, options);
      if (!result.success || this.until(result)) {
        return result;
      }
      current = outputToInput(current, result.output);
    }
  }
}

export class BranchNode extends Node {
  constructor(private readonly selector: BranchSelector) {
    super();
  }

  override async run(input: PipelineInput, options: EngineOptions): Promise<PipelineResult> {
    return runPipelineItem(this.selector(input), input, options);
  }
}
